### ISSUE TITLE
SCIM API for SAML linked accounts

### ISSUE THREAD

**Issue by isg-ncl-kd** [PyGithub/PyGithub#1448]
We would like to be able to make use of the SCIM API (https://developer.github.com/v3/scim/) within the pyGithub package. This would enable us to search users in our organisation by their SCIM attributes, such as the idp provided email address, nameID. These attributes cannot be changed by the organisation members, unlike their personal github account details, so would provide an immutable set of search criteria for us.

**Comment by fl-max** [comment]
Just a note, this won't work if SSO is set at the Enterprise level. You'll instead need to use a GraphQL query. An example can be found [here](https://github.com/github/platform-samples/blob/master/graphql/queries/enterprise-sso-member-details.graphql).

Using a session with the proper auth headers, you'd send this query like so:
```python
res = session.post("https://api.github.com/graphql", json={'query': query})
```

Where query is the GraphQL query. Loading the query from a file and using Jinja2 to template it helps. This does require admin:Org and admin:Enterprise permissions to work.

**Comment by anotherbhav** [comment]
I had a little trouble gluing all the information out there together, but this works for me to get a list of mapping dicts with 2 values: the sso email (in most cases) and the GitHub handle.  This is is the only way I could figure out how. 

The SCIM documentation does not provide the GitHub handle or unique identifier that is also on the Org User list.  Hope it helps!

The PyGitHub had the functionality for graphql queries, but there wasn't much documentation on how to use it (that I could find), but it was laid out in an understandable way (Thanks team!).

```python
from github import Github
from github import Auth
import json

api_token = "ghp_secrett0ken234"
auth = Auth.Token(api_token)
gh_conn = Github(auth=auth)

org_name = "MyOrg"
end_cursor = None  # First end_cursor must be null to get the first page
users = []

while [True]:
    resp = gh_conn.requester.graphql_query("""
    query ($orgName: String!, $ssoCursor: String) {
        organization(login: $orgName) {
            samlIdentityProvider {
                ssoUrl
                externalIdentities(after: $ssoCursor, first: 100) {
                    totalCount
                    pageInfo {
                        hasNextPage
                        endCursor
                    }
                    edges {
                        node {
                            user {
                                login
                            }
                            scimIdentity {
                                username
                            }
                        }
                    }
                }
            }
        }
    }
                                    """, variables={"orgName": org_name, "ssoCursor" : end_cursor})

    data = resp[1] # the data is in the second position of the list
    users += data['data']['organization']['samlIdentityProvider']['externalIdentities']['edges']
    
    page_info = data['data']['organization']['samlIdentityProvider']['externalIdentities']['pageInfo']
    if 'hasNextPage' in page_info and page_info['hasNextPage']:
        end_cursor = page_info['endCursor']
    else:
        break # No end cursor found, end loop
    
gh_conn.close()
```

## references
- https://github.com/github/platform-samples/issues/168
- https://stackoverflow.com/questions/50589221/access-linked-saml-identity-via-the-github-api

**Comment by cyberious** [comment]
This is now available via the `https://api.github.com/scim/v2/enterprises/<ENT>/Users?filter=username eq "username"` we should move this back up as it is relevant. It is far less costly of a api call and organization would be no longer needed.

**Comment by EnricoMi** [comment]
I agree, given this is available through a REST API, that one should be preferred over using GraphQL.

The docs can be found here:
https://docs.github.com/de/enterprise-cloud@latest/rest/enterprise-admin/scim?apiVersion=2022-11-28#list-scim-provisioned-identities-for-an-enterprise
